[GeoMechanics] Added Triangular element 15 nodes, Line element 5 nodes

mnabideltares opened the pull request:
**📝 Description**
Added Triangular geometry 15 nodes:

-Triangle2D15N
- Line2D5N

**🆕 Changelog**
The Geometry is added to generate elements. 
- TransientPwElement2D15N
- SteadyStatePwElement2D15N
- UPwSmallStrainElement2D15N
- SmallStrainUPwDiffOrderElement2D15N
- UPwSmallStrainAxisymmetricElement2D15N
- SmallStrainUPwDiffOrderAxisymmetricElement2D15N
- UPwUpdatedLagrangianElement2D15N
- UpdatedLagrangianUPwDiffOrderElement2D15N
- UPwUpdatedLagrangianAxisymmetricElement2D15N
- UpdatedLagrangianUPwDiffOrderAxisymmetricElement2D15N
- LineLoadDiffOrderCondition2D5N
- LineNormalLoadDiffOrderCondition2D5N
- LineNormalFluidFluxDiffOrderCondition2D5N
-

mnabideltares committed:
Added Triangular element 15 nodes, Line element 5 nodes

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_conditions/general_U_Pw_diff_order_condition.cpp:
```diff
@@ -75,6 +75,9 @@ void GeneralUPwDiffOrderCondition::Initialize(const ProcessInfo& rCurrentProcess
         case 3: //2D L3P2
             mpPressureGeometry = GeometryType::Pointer( new Line2D2< Node<3> >(rGeom(0), rGeom(1)) );
             break;
+        case 5: //2D L5P2
```
I think, in U-Pw different order elements/conditions, we use one order lower shape functions for pore pressures. Here, 2 orders lower is used. Is it the intended behaviour?

mnabideltares replied:
Right, used 2 order less, because:
1. I didn't commit 1 order less, namely Triangle2D10N, at moment, because to avoid a large commit. Will be added in the next commit.
2. As far as I understand, linear order for pressure leads to higher satiability.  So, it was intended behavior.

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp:
```diff
@@ -259,7 +259,28 @@ void UPwBaseElement<TDim,TNumNodes>::
 GeometryData::IntegrationMethod UPwBaseElement<TDim,TNumNodes>::GetIntegrationMethod() const
 {
-    return GeometryData::IntegrationMethod::GI_GAUSS_2;
+    GeometryData::IntegrationMethod GI_GAUSS;
```
I do not get why this is changed. GI_GAUSS_2 is not enough?

mnabideltares replied:
GI_GAUSS_2 does not let the model converge for 15N element. Needs to add the number of integration points.

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+//    |  /           |
```
Perhaps, this file can be moved to a separate directory, for example "custom_geomteries"

mnabideltares replied:
I am requested to move these element to the Kratos core. See #10083.

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+            array_1d<double, 3> current_global_coords;
+
+            //Newton iteration:
+            for (IndexType k = 0; k < MaxIteratioNumberPointLocalCoordinates; k++) {
```
Minor comment: using ++k instead of k++ is better in for loops.

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+        {
+            if (rResult.size() != 5) rResult.resize(5, false);
+            //
+            double xcor = rCoordinates[0];
```
Please use "const" if the variable will not change later.

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+         * @see ShapeFunctionsLocalGradients
+         * @see ShapeFunctionLocalGradient
+         */
+        double ShapeFunctionValue(IndexType ShapeFunctionIndex, const CoordinatesArrayType& rPoint) const override
```
Shouldn't be const IndexType ShapeFunctionIndex?

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+         */
+        double ShapeFunctionValue(IndexType ShapeFunctionIndex, const CoordinatesArrayType& rPoint) const override
+        {
+            double shape = 0.0;
```
Please use "const" in front of double if the variables will not change later.

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+        // ========================================================================================
+        /** Calculates the local gradients for all integration points for given integration method
+         */
+        virtual ShapeFunctionsGradientsType ShapeFunctionsLocalGradients(IntegrationMethod ThisMethod)
```
It seems that a copy is made here. Shouldn't be "const IntegrationMethod &ThisMethod" instead of "IntegrationMethod ThisMethod"?

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+            const int integration_points_number = msGeometryData.IntegrationPointsNumber(ThisMethod);
+            ShapeFunctionsGradientsType Result(integration_points_number);
+            //
+            for (int pnt = 0; pnt < integration_points_number; pnt++)
```
++pnt

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+            if (rResult.size1() != 5 || rResult.size2() != 1) rResult.resize(5, 1, false);
+            noalias(rResult) = ZeroMatrix(5, 1);
+            //
+            double xcor = rPoint[0];
```
const

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+         * shape functions in given point
+         * @param rPoint the given point the gradients are calculated in
+         */
+        virtual Matrix& ShapeFunctionsGradients(Matrix& rResult, CoordinatesArrayType& rPoint)
```
const CoordinatesArrayType& rPoint?

Vahid-Galavi replied:
It seems that rPoint is constant in this function.

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+            if (rResult.size1() != 5 || rResult.size2() != 1) rResult.resize(5, 1, false);
+            noalias(rResult) = ZeroMatrix(5, 1);
+            //
+            double xcor = rPoint[0];
```
const

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1314 @@
+            int integration_points_number = IntegrationPoints.size();
+            Matrix shape_function_values(integration_points_number, 5);
+            //
+            double xcor = 0.0;
```
use more const, if possible.

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:
```diff
@@ -876,6 +879,25 @@ void SmallStrainUPwDiffOrderElement::AssignPressureToIntermediateNodes()
             ThreadSafeNodeWrite(rGeom[9],WATER_PRESSURE, 0.5 * (p2 + p3) );
             break;
         }
+        case 15: //2D T15P3
```
OK, You are always using a linear interpolation for water flow which is good for partially saturated flow.

mnabideltares replied:
Right, See the comment above

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+#pragma once
```
All the files related to geometry should be moved to a different directory, for example "custom_geometries"

mnabideltares replied:
See the comment above. They will be moved to Kratos core

Vahid-Galavi reviewed (commented):
I will do the rest of review, later

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+        {
+            if (rResult.size() != 15) rResult.resize(15, false);
+            //
+            double cof1 = 128.0 / 3.0;
```
use const for these variables

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+        double ShapeFunctionValue(IndexType ShapeFunctionIndex,
+            const CoordinatesArrayType& rPoint) const override
+        {
+            double cof1 = 128.0 / 3.0;
```
const

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+            switch (ShapeFunctionIndex) {
+            case 0:
+                shape = zta * (zta - 0.25) * (zta - 0.5) * (zta - 0.75) * cof2;
+                return(shape);
```
I  think parentheses can be removed.

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+                NodesInFaces.resize(6, 3, false);
+
+            //compatible to Edges() function. Triangle library considers a different order
+            NodesInFaces(0, 0) = 0;//face or master node
```
it is not compatible with the ordering mentioned in this issue: [https://github.com/KratosMultiphysics/Kratos/issues/9990](url)

mnabideltares replied:
Still not clear to me. I see the order is different than the order in triangular element, as it is done for 2D6N. I will investigate it in mode details.

Vahid-Galavi replied:
I just meant that the ordering that you mentioned in the above link is different from what is implemented here. Maybe you used another source. This was just a minor remark.

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+            const int integration_points_number = msGeometryData.IntegrationPointsNumber(ThisMethod);
+            ShapeFunctionsGradientsType Result(integration_points_number);
+            //
+            for (int pnt = 0; pnt < integration_points_number; pnt++)
```
++pnt

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+        Matrix& ShapeFunctionsLocalGradients(Matrix& rResult, const CoordinatesArrayType& rPoint) const override
+        {
+            rResult.resize(15, 2, false);
+            double x1 = rPoint[0];
```
use const as much as possible

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+         * shape functions in given point
+         * @param rPoint the given point the gradients are calculated in
+         */
+        virtual Matrix& ShapeFunctionsGradients(Matrix& rResult, CoordinatesArrayType& rPoint)
```
shouldn't be "const CoordinatesArrayType& rPoint"?

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+        virtual Matrix& ShapeFunctionsGradients(Matrix& rResult, CoordinatesArrayType& rPoint)
+        {
+            rResult.resize(15, 2, false);
+            double x1 = rPoint[0];
```
use const as much as possible

Vahid-Galavi commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+                ShapeFunctionsGradientsType temp(this->PointsNumber());
+                rResult.swap(temp);
+            }
+            double x1 = rPoint[0];
```
const

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -65,6 +67,7 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mTransientPwElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mTransientPwElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mTransientPwElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mTransientPwElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mTransientPwElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -82,6 +85,7 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mSteadyStatePwElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mSteadyStatePwElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mSteadyStatePwElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mSteadyStatePwElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mSteadyStatePwElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -99,6 +103,7 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUPwSmallStrainElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUPwSmallStrainElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUPwSmallStrainElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mUPwSmallStrainElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mUPwSmallStrainElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -125,6 +130,7 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mSmallStrainUPwDiffOrderElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mSmallStrainUPwDiffOrderElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mSmallStrainUPwDiffOrderElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mSmallStrainUPwDiffOrderElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mSmallStrainUPwDiffOrderElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -135,11 +141,13 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUPwSmallStrainAxisymmetricElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUPwSmallStrainAxisymmetricElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUPwSmallStrainAxisymmetricElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mUPwSmallStrainAxisymmetricElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mUPwSmallStrainAxisymmetricElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -135,11 +141,13 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mSmallStrainUPwDiffOrderAxisymmetricElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mSmallStrainUPwDiffOrderAxisymmetricElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mSmallStrainUPwDiffOrderAxisymmetricElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mSmallStrainUPwDiffOrderAxisymmetricElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mSmallStrainUPwDiffOrderAxisymmetricElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -159,6 +167,7 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUPwUpdatedLagrangianElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUPwUpdatedLagrangianElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUPwUpdatedLagrangianElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mUPwUpdatedLagrangianElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mUPwUpdatedLagrangianElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -171,6 +180,7 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUpdatedLagrangianUPwDiffOrderElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUpdatedLagrangianUPwDiffOrderElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUpdatedLagrangianUPwDiffOrderElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mUpdatedLagrangianUPwDiffOrderElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mUpdatedLagrangianUPwDiffOrderElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -181,9 +191,11 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUPwUpdatedLagrangianAxisymmetricElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUPwUpdatedLagrangianAxisymmetricElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUPwUpdatedLagrangianAxisymmetricElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+    mUPwUpdatedLagrangianAxisymmetricElement2D15N(0, Element::GeometryType::Pointer(new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```
```suggestion
    mUPwUpdatedLagrangianAxisymmetricElement2D15N(0, Element::GeometryType::Pointer( new Triangle2D15 <NodeType>(Element::GeometryType::PointsArrayType(15)))),
```

mnabideltares replied:
Done

Vahid-Galavi reviewed (commented):
I went through all the files

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1753 @@
+     *          |   `\
+     *          10 14 7
+     *          |      '\
+     *          11 12 13 6
```
Great!! What was the problem a few weeks ago that prevented this element to work? :)

mnabideltares replied:
Thank you @AlejandroCornejo ,
In the Geo core, the number of Gauss points is hard programmed to use Gauss_2 (3 points) for specific type of elements. The 15 nodes element was not converging because of this limitation.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1109,7 +1109,7 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
     //Previous definitions
     const PropertiesType& rProp = this->GetProperties();
     const GeometryType& rGeom = this->GetGeometry();
```
I do not know the whole code style in the GeoApp but this should be `r_geom` in the Structural for example :D

mnabideltares replied:
See #10083 (Migrate the new designed elements from GeoMechanics branch to Kratos application).
I will move them to the Kratos core after that this issue is reviewed.

mnabideltares replied:
rGeom is used as the default variable name in GroMechanics branch

mnabideltares committed:
Revision 1: Added Triangular element 15 nodes, Line element 5 nodes

Adjustment based on the review requirements.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_face_load_condition.cpp:
```diff
@@ -107,6 +107,21 @@ void UPwFaceLoadCondition<2,3>::
+    const Matrix& Jacobian,
+    const double& Weight)
+{
+    double dx_dxi = Jacobian(0, 0);
```
const?

mnabideltares replied:
Thank you @AlejandroCornejo for your help. 
1. Related to your comment, I am working to add them to the Kratos core, and I am adding test cases too. It will be committed during this week, if everything goes smoothly.
 2. const is added

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_face_load_condition.cpp:
```diff
@@ -107,6 +107,21 @@ void UPwFaceLoadCondition<2,3>::
+    double dx_dxi = Jacobian(0, 0);
+    double dy_dxi = Jacobian(1, 0);
+
+    double ds = sqrt(dx_dxi * dx_dxi + dy_dxi * dy_dxi);
```
const

mnabideltares replied:
Added

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+        }
+
+        // ========================================================================================
+        // ========================================================================================
```
I think that this style is not consistent with the rest of geometries right?

mnabideltares replied:
I added such lines in order to make it visually easier to distinguish the start of each function. However, in order to be consistent with the rest of code, do I remove these lines?

AlejandroCornejo replied:
it is always better to maintain the code style according to the previous ones :)

mnabideltares replied:
Adjusted

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+
+            static constexpr double MaxNormPointLocalCoordinates = 300.0;
+            static constexpr std::size_t MaxIteratioNumberPointLocalCoordinates = 500;
+            static constexpr double MaxTolerancePointLocalCoordinates = 1.0e-8;
```
this ad hoc settings maybe it is better to put them as a default in the method input right?

mnabideltares replied:
it is override method. If I change here, I have to change the main function too. Then I need to change for all other elements, I guess. 
So, it is better to let it stay temporarily as an ad hoc solution. Do you agree?

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+            // Starting with xi = 0
+            if (rResult.size() != 5)
+                rResult.resize(5, false);
+            rResult = ZeroVector(5);
```
noalias(rResult) = ...

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+                noalias(DN) = prod(X, shape_functions_gradients);
+
+                noalias(J) = prod(trans(DN), DN);
+                const array_1d<double, 1> res = prod(trans(DN), current_global_coords);
```
I think that it is more efficient to declare the res vector and then use noalias

mnabideltares replied:
Done.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+                = CalculateShapeFunctionsIntegrationPointsLocalGradients(ThisMethod);
+            Matrix shape_function_gradient_in_integration_point = shape_functions_gradients(IntegrationPointIndex);
+            // Values of shape functions in integration points
+            DenseVector<double> ShapeFunctionsValuesInIntegrationPoint = ZeroVector(3);
```
why not using Vector?

mnabideltares replied:
Don't know, I copied it from line_2d_3.h
Should I change it to Vector?

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+        ///@name Shape Function Integration Points Gradient
+        ///@{
+        void ShapeFunctionsIntegrationPointsGradients(ShapeFunctionsGradientsType& rResult,
+            IntegrationMethod ThisMethod) const override
```
IntegrationMethodd&

mnabideltares replied:
Done

mnabideltares replied:
Get error, so I removed the "&" :

D:\temp04\Kratos\applications\GeoMechanicsApplication\custom_elements/line_2d_5.hpp(772,14): error C3668: 'Kratos::Line2D5<TPointType>::ShapeFunctionsIntegrationPointsGradients': method with override specifier 'override' did not override any base class methods [D:\temp04\Kratos\build\Debug\applications\GeoMechanicsApplication\KratosGeoMechanicsCore.vcxpro
j]

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+            rResult(2, 0) = -(gx1 * (gx3 - fx3 - 1.0) + 2.0 * xi * fx3) * cof2;
+            rResult(3, 0) = 2.0 * xi * (gx3 + 4.0 * fx3);
+            rResult(4, 0) = -(gx2 * (gx3 - fx3 - 1.0) + 2.0 * xi * fx3) * cof2;
+            return(rResult);
```
why the ( )?

mnabideltares replied:
Removed ( )

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+            {
+                double xi = IntegrationPoints[pnt].X();
+                double fx1 = xi - 1.0;
+                double fx2 = xi + 1.0;
```
const :)

mnabideltares replied:
They are in a foor loop, and change for each integration point

Vahid-Galavi replied:
But still, you can use const. Because you do not change them after the initialisation.

mnabideltares replied:
Set const in the loop. Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+            const IntegrationPointsContainerType& all_integration_points = AllIntegrationPoints();
+            const IntegrationPointsArrayType& IntegrationPoints
+                = all_integration_points[static_cast<int>(ThisMethod)];
+            int integration_points_number = IntegrationPoints.size();
```
const int

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/line_2d_5.hpp:
```diff
@@ -0,0 +1,1301 @@
+            {
+                double xi = IntegrationPoints[pnt].X();
+                double fx1 = xi - 1.0;
+                double fx2 = xi + 1.0;
```
const

mnabideltares replied:
In a for loop

Vahid-Galavi replied:
You can use `const` in a for loop.

mnabideltares replied:
Set const in the loop. Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+            const PointType& Point13, const PointType& Point14, const PointType& Point15)
+            : BaseType(PointsArrayType(), &msGeometryData)
+        {
+            this->Points().push_back(typename PointType::Pointer(new PointType(Point01)));
```
can't we create an aux variable called `auto& r_points = this->Points()` and then `r_points .push_back` it avoids calling all the time the this->Points()

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+        Matrix& PointsLocalCoordinates(Matrix& rResult) const override
+        {
+            rResult.resize(15, 2, false);
+            noalias(rResult) = ZeroMatrix(15, 2);
```
actually if you're stting all the values this line can be avoided

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+         */
+        double Length() const override
+        {
+            return sqrt(fabs(Area()));
```
`std::sqrt(std::abs(Area()))`

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+            const double zt3 = zt2 * zt1;
+
+            //// noalias(rResult) = ZeroMatrix(15, 2);
+            rResult = ZeroMatrix(15, 2);
```
noalias, maybe this line is unnecessary

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+                ShapeFunctionsThirdDerivativesType temp(this->PointsNumber());
+                rResult.swap(temp);
+            }
+            for (IndexType i = 0; i < rResult.size(); i++)
```
++i

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+            rResult[14][1].resize(2, 2, false);
+            //
+            double fx3 = 256.0 * rPoint[0] + 256.0 * rPoint[1] - 160.0;
+            double fx2y = fx3;
```
const?

mnabideltares replied:
These variables are re-used in the same method (overwritten). They are not constants.

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+            const int integration_points_number = integration_points.size();
+            ShapeFunctionsGradientsType d_shape_f_values(integration_points_number);
+            //initialising container
+            //std::fill(d_shape_f_values.begin(), d_shape_f_values.end(), Matrix(4,2));
```
remove unused code?

mnabideltares replied:
Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+            {
+                Matrix result(15, 2);
+                double xi1 = integration_points[pnt].X();
+                double xi2 = xi1 * xi1;
```
const?

mnabideltares replied:
These variables are in for loop and they are overwritten for each integration point.

mnabideltares replied:
Set const in the loop, works fine. Done

AlejandroCornejo commented on applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp:
```diff
@@ -0,0 +1,1752 @@
+                double zt3 = zt2 * zt1;
+
+                //// noalias(result) = ZeroMatrix(15, 2);
+                result = ZeroMatrix(15, 2);
```
noalias or remove even I think

mnabideltares replied:
noalias is added

AlejandroCornejo reviewed (changes requested):
Besides my comments, I think that a set of tests must be added, specially for the geometries that have been implemented to the core (line and triangle). In this way nobody will break your precious code :)

Vahid-Galavi reviewed (dismissed):
I approve the changes.

mnabideltares committed:
Revision 2: Added Triangular element 15 nodes, Line element 5 nodes

Adjustment based on the review requirements.

Vahid-Galavi reviewed (commented):
I have no comment anymore. Let's wait for @AlejandroCornejo

Vahid-Galavi reviewed (dismissed):
Great!

mnabideltares committed:
Merge branch 'master' into 9990-geomechanicsapplication-add-15-node-triangular-elements

mnabideltares committed:
Added 10 node triangle elements, and 4 nodes line element

NB: small_strain_U_Pw_diff_order_element.cpp is still need to be modified to account for pressure ellement. The initialization and interpolations are still incorrect.

mnabideltares committed:
Merge branch 'master' into 9990-geomechanicsapplication-add-15-node-triangular-elements

mnabideltares committed:
for different order element, definition and interpolations are defined for pressure, based on one order lower than the deformation element

For teriangles with 10 and 15 nodes

AlejandroCornejo reviewed (dismissed):
This element already has a test right?

mnabideltares committed:
Test cases for elements Triangle2D10N and Triangle2D15N are added

mnabideltares committed:
Minor fix, A white space is removed for code quality check

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -165,6 +180,8 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUPwUpdatedLagrangianElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUPwUpdatedLagrangianElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUPwUpdatedLagrangianElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+	mUPwUpdatedLagrangianElement2D10N( 0, Element::GeometryType::Pointer( new Triangle2D10 <NodeType >( Element::GeometryType::PointsArrayType(10)))),
```
there are some spaces in the beginning of the line

mnabideltares replied:
@Vahid-Galavi  Thank you for reviewing this work, very appreciated.

Removed the space. 
These spaces were defined as tabs (by VS IDE) as VS was not showing them.

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -177,6 +194,8 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUpdatedLagrangianUPwDiffOrderElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUpdatedLagrangianUPwDiffOrderElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUpdatedLagrangianUPwDiffOrderElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+	mUpdatedLagrangianUPwDiffOrderElement2D10N( 0, Element::GeometryType::Pointer( new Triangle2D10 <NodeType >( Element::GeometryType::PointsArrayType(10)))),
```
the same

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -187,9 +206,13 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D6N( 0, Element::GeometryType::Pointer( new Triangle2D6 <NodeType >( Element::GeometryType::PointsArrayType(6)))),
     mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D8N( 0, Element::GeometryType::Pointer( new Quadrilateral2D8 <NodeType >( Element::GeometryType::PointsArrayType(8)))),
     mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D9N( 0, Element::GeometryType::Pointer( new Quadrilateral2D9 <NodeType >( Element::GeometryType::PointsArrayType(9)))),
+	mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D10N( 0, Element::GeometryType::Pointer( new Triangle2D10 <NodeType >( Element::GeometryType::PointsArrayType(10)))),
```
the same

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -223,6 +246,8 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
 
     mUPwFaceLoadCondition2D3N( 0, Condition::GeometryType::Pointer( new Line2D3<NodeType >( Condition::GeometryType::PointsArrayType(3)))),
+	mUPwFaceLoadCondition2D4N( 0, Condition::GeometryType::Pointer( new Line2D4<NodeType >( Condition::GeometryType::PointsArrayType(4)))),
+	mUPwFaceLoadCondition2D5N( 0, Condition::GeometryType::Pointer( new Line2D5<NodeType >( Condition::GeometryType::PointsArrayType(5)))),
```
remove the spaces

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -234,8 +259,14 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
     mUPwNormalFluxFICCondition3D4N( 0, Condition::GeometryType::Pointer( new Quadrilateral3D4 <NodeType >( Condition::GeometryType::PointsArrayType(4)))),
 
     mLineLoadDiffOrderCondition2D3N( 0, Condition::GeometryType::Pointer( new Line2D3<NodeType >( Condition::GeometryType::PointsArrayType(3)))),
+	mLineLoadDiffOrderCondition2D4N( 0, Condition::GeometryType::Pointer( new Line2D4<NodeType >( Condition::GeometryType::PointsArrayType(4)))),
```
space

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -234,8 +259,14 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
+	mLineLoadDiffOrderCondition2D4N( 0, Condition::GeometryType::Pointer( new Line2D4<NodeType >( Condition::GeometryType::PointsArrayType(4)))),
+    mLineLoadDiffOrderCondition2D5N( 0, Condition::GeometryType::Pointer( new Line2D5<NodeType >( Condition::GeometryType::PointsArrayType(5)))),
     mLineNormalLoadDiffOrderCondition2D3N( 0, Condition::GeometryType::Pointer( new Line2D3<NodeType >( Condition::GeometryType::PointsArrayType(3)))),
+	mLineNormalLoadDiffOrderCondition2D4N( 0, Condition::GeometryType::Pointer( new Line2D4<NodeType >( Condition::GeometryType::PointsArrayType(4)))),
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -234,8 +259,14 @@ KratosGeoMechanicsApplication::KratosGeoMechanicsApplication():
+	mLineNormalLoadDiffOrderCondition2D4N( 0, Condition::GeometryType::Pointer( new Line2D4<NodeType >( Condition::GeometryType::PointsArrayType(4)))),
+    mLineNormalLoadDiffOrderCondition2D5N( 0, Condition::GeometryType::Pointer( new Line2D5<NodeType >( Condition::GeometryType::PointsArrayType(5)))),
     mLineNormalFluidFluxDiffOrderCondition2D3N( 0, Condition::GeometryType::Pointer( new Line2D3<NodeType >( Condition::GeometryType::PointsArrayType(3)))),
+	mLineNormalFluidFluxDiffOrderCondition2D4N( 0, Condition::GeometryType::Pointer( new Line2D4<NodeType >( Condition::GeometryType::PointsArrayType(4)))),
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -295,6 +328,8 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "SteadyStatePwElement2D6N", mSteadyStatePwElement2D6N )
     KRATOS_REGISTER_ELEMENT( "SteadyStatePwElement2D8N", mSteadyStatePwElement2D8N )
     KRATOS_REGISTER_ELEMENT( "SteadyStatePwElement2D9N", mSteadyStatePwElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "SteadyStatePwElement2D10N", mSteadyStatePwElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -316,6 +351,8 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "UPwSmallStrainElement2D6N", mUPwSmallStrainElement2D6N )
     KRATOS_REGISTER_ELEMENT( "UPwSmallStrainElement2D8N", mUPwSmallStrainElement2D8N )
     KRATOS_REGISTER_ELEMENT( "UPwSmallStrainElement2D9N", mUPwSmallStrainElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "UPwSmallStrainElement2D10N", mUPwSmallStrainElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -342,6 +379,8 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderElement2D6N", mSmallStrainUPwDiffOrderElement2D6N )
     KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderElement2D8N", mSmallStrainUPwDiffOrderElement2D8N )
     KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderElement2D9N", mSmallStrainUPwDiffOrderElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderElement2D10N", mSmallStrainUPwDiffOrderElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -352,13 +391,17 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "UPwSmallStrainAxisymmetricElement2D6N", mUPwSmallStrainAxisymmetricElement2D6N )
     KRATOS_REGISTER_ELEMENT( "UPwSmallStrainAxisymmetricElement2D8N", mUPwSmallStrainAxisymmetricElement2D8N )
     KRATOS_REGISTER_ELEMENT( "UPwSmallStrainAxisymmetricElement2D9N", mUPwSmallStrainAxisymmetricElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "UPwSmallStrainAxisymmetricElement2D10N", mUPwSmallStrainAxisymmetricElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -352,13 +391,17 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderAxisymmetricElement2D6N", mSmallStrainUPwDiffOrderAxisymmetricElement2D6N )
     KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderAxisymmetricElement2D8N", mSmallStrainUPwDiffOrderAxisymmetricElement2D8N )
     KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderAxisymmetricElement2D9N", mSmallStrainUPwDiffOrderAxisymmetricElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "SmallStrainUPwDiffOrderAxisymmetricElement2D10N", mSmallStrainUPwDiffOrderAxisymmetricElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -378,6 +421,8 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianElement2D6N", mUPwUpdatedLagrangianElement2D6N )
     KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianElement2D8N", mUPwUpdatedLagrangianElement2D8N )
     KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianElement2D9N", mUPwUpdatedLagrangianElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianElement2D10N", mUPwUpdatedLagrangianElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -390,6 +435,8 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderElement2D6N", mUpdatedLagrangianUPwDiffOrderElement2D6N )
     KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderElement2D8N", mUpdatedLagrangianUPwDiffOrderElement2D8N )
     KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderElement2D9N", mUpdatedLagrangianUPwDiffOrderElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderElement2D10N", mUpdatedLagrangianUPwDiffOrderElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -400,10 +447,14 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianAxisymmetricElement2D6N", mUPwUpdatedLagrangianAxisymmetricElement2D6N )
     KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianAxisymmetricElement2D8N", mUPwUpdatedLagrangianAxisymmetricElement2D8N )
     KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianAxisymmetricElement2D9N", mUPwUpdatedLagrangianAxisymmetricElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "UPwUpdatedLagrangianAxisymmetricElement2D10N", mUPwUpdatedLagrangianAxisymmetricElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -400,10 +447,14 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderAxisymmetricElement2D6N", mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D6N )
     KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderAxisymmetricElement2D8N", mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D8N )
     KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderAxisymmetricElement2D9N", mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D9N )
+	KRATOS_REGISTER_ELEMENT( "UpdatedLagrangianUPwDiffOrderAxisymmetricElement2D10N", mUpdatedLagrangianUPwDiffOrderAxisymmetricElement2D10N )
```
space

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -449,8 +502,14 @@ void KratosGeoMechanicsApplication::Register() {
     KRATOS_REGISTER_CONDITION( "UPwNormalFluxFICCondition3D4N", mUPwNormalFluxFICCondition3D4N )
 
     KRATOS_REGISTER_CONDITION( "LineLoadDiffOrderCondition2D3N", mLineLoadDiffOrderCondition2D3N )
+	KRATOS_REGISTER_CONDITION( "LineLoadDiffOrderCondition2D4N", mLineLoadDiffOrderCondition2D4N )
```
space

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -449,8 +502,14 @@ void KratosGeoMechanicsApplication::Register() {
+	KRATOS_REGISTER_CONDITION( "LineLoadDiffOrderCondition2D4N", mLineLoadDiffOrderCondition2D4N )
+    KRATOS_REGISTER_CONDITION( "LineLoadDiffOrderCondition2D5N", mLineLoadDiffOrderCondition2D5N )
     KRATOS_REGISTER_CONDITION( "LineNormalLoadDiffOrderCondition2D3N", mLineNormalLoadDiffOrderCondition2D3N )
+	KRATOS_REGISTER_CONDITION( "LineNormalLoadDiffOrderCondition2D4N", mLineNormalLoadDiffOrderCondition2D4N )
```
space

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -449,8 +502,14 @@ void KratosGeoMechanicsApplication::Register() {
+	KRATOS_REGISTER_CONDITION( "LineNormalLoadDiffOrderCondition2D4N", mLineNormalLoadDiffOrderCondition2D4N )
+    KRATOS_REGISTER_CONDITION( "LineNormalLoadDiffOrderCondition2D5N", mLineNormalLoadDiffOrderCondition2D5N )
     KRATOS_REGISTER_CONDITION( "LineNormalFluidFluxDiffOrderCondition2D3N", mLineNormalFluidFluxDiffOrderCondition2D3N )
+	KRATOS_REGISTER_CONDITION( "LineNormalFluidFluxDiffOrderCondition2D4N", mLineNormalFluidFluxDiffOrderCondition2D4N )
```
space

mnabideltares replied:
Done

Vahid-Galavi commented on applications/GeoMechanicsApplication/tests/test_triangle_10n.gid/test_triangle_10n.mdpa:
```diff
@@ -0,0 +1,207 @@
+End Nodes
+
+
+Begin Elements SmallStrainUPwDiffOrderElement2D10N
```
is it possible to reduce number of elements in this test to reduce the size of mdpa file?

mnabideltares replied:
Reduced the number of elements from 8 to 4.

Vahid-Galavi commented on applications/GeoMechanicsApplication/tests/test_triangle_15n.gid/test_triangle_15n.mdpa:
```diff
@@ -0,0 +1,279 @@
+End Nodes
+
+
+Begin Elements SmallStrainUPwDiffOrderElement2D15N
```
can you reduce number of elements to reduce the size of the mdpa file?

mnabideltares replied:
Reduced the number of elements from 8 to 4.

Vahid-Galavi reviewed (commented):
I only have some minor remarks.

mnabideltares committed:
Fixes based or review request
